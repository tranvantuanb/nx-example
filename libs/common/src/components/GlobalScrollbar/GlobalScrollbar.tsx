import React, { useState, useLayoutEffect, useEffect } from 'react';
import { Global } from '@emotion/react';
import { css } from 'twin.macro';

// https://github.com/react-component/overflow/issues/6
const useIsomorphicLayoutEffect =
  typeof window !== 'undefined' ? useLayoutEffect : useEffect;

export interface GlobalScrollbarProps {
  className?: string;
}

/**
 * Customize website's scrollbar like Mac OS only in other's OS.
 * The customization is not applied in Mac OS.
 */
export const GlobalScrollbar: React.FC<GlobalScrollbarProps> = React.memo(
  ({ className }) => {
    const [isApple, setIsApple] = useState(true);

    useIsomorphicLayoutEffect(() => {
      const isNotApple = !/(Mac|iPhone|iPod|iPad)/i.test(navigator.platform);
      if (isNotApple) {
        setIsApple(false);
      }
    }, []);

    return isApple ? null : (
      <Global
        styles={css`
          /* total width */
          ::-webkit-scrollbar {
            background-color: #fff;
            width: 16px;
          }

          /* background of the scrollbar except button or resizer */
          ::-webkit-scrollbar-track {
            background-color: #fff;
          }

          /* scrollbar itself */
          ::-webkit-scrollbar-thumb {
            background-color: #babac0;
            border-radius: 16px;
            border: 4px solid #fff;
          }

          /* set button(top and bottom of the scrollbar) */
          ::-webkit-scrollbar-button {
            display: none;
          }
        `}
      />
    );
  }
);
