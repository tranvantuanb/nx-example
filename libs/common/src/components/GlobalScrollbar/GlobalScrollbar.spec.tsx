import { render } from '../../test';

import { GlobalScrollbar, GlobalScrollbarProps } from './GlobalScrollbar';

describe('GlobalScrollbar', () => {
  it('should render successfully', () => {
    const defaultProps: GlobalScrollbarProps = {};
    const { container } = render(<GlobalScrollbar {...defaultProps} />);
    expect(container).toBeTruthy();
    expect(container).toMatchSnapshot();
  });
});
