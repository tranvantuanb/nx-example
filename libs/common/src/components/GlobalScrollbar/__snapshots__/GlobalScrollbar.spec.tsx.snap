// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalScrollbar should render successfully 1`] = `<div />`;
